### PR TITLE
Add data URL support

### PR DESCRIPTION
**Pull request by fisker**
fixes: #94  & #84 

questions:

1. url `data:text/plain;charset=UTF-8;,foo` mentioned in #84, is not valid url accounding to `valid-data-url`, should I change Regex to loose strict one?
2. any more test cases need cover?
3. are we going to support other protocols like `file:` `ftp:` etc

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#94: Normalize data URLs](https://issuehunt.io/repos/29094022/issues/94)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -6,6 +6,52 @@ const testParameter = (name, filters) => {
 	return filters.some(filter => filter instanceof RegExp ? filter.test(name) : filter === name);
 };
 
+// From https://github.com/killmenot/valid-data-url/blob/8d56cab866469a4608001f0e8c5d73f65789ba13/index.js#L24
+// eslint-disable-next-line no-useless-escape
```
Comment (sindresorhus) [inline index.js]: If ESLint reports this, it's probably correct and you should remove the useless escape.

File: index.js [hunk 2]
```diff
@@ -41,7 +83,7 @@ const normalizeUrl = (urlString, options) => {
 	const isRelativeUrl = !hasRelativeProtocol && /^\.*\//.test(urlString);
 
 	// Prepend protocol
-	if (!isRelativeUrl) {
+	if (!isRelativeUrl && !/^data?:/i.test(urlString)) {
```
Comment (sindresorhus) [inline index.js]: Why the `?`

**Comment by sindresorhus** [comment]
>  url data:text/plain;charset=UTF-8;,foo mentioned in #84, is not valid url accounding to valid-data-url, should I change Regex to loose strict one?

Yes, but make sure it's a valid data URL according to the spec first: https://tools.ietf.org/html/rfc2397

> any more test cases need cover?

I can't really think of more cases.

> are we going to support other protocols like file: ftp: etc

Would be useful to support `file:` as it's quite common. Not interested in supporting `ftp` though.

**Comment by sindresorhus** [comment]
Would be good to mention data URL support in the readme and index.d.ts. I don't think it's obvious.

File: index.js (hunk 2 above)
Comment (fisker) [reply index.js]: Mistake. fixed

**Comment by fisker** [comment]
> Would be good to mention data URL support in the readme and index.d.ts. I don't think it's obvious.

Not sure how do you want to do this, add an example? or explain `url` parameter?

**Comment by sindresorhus** [comment]
> or explain url parameter?

This

**Comment by fisker** [comment]
@sindresorhus ready to review

**Comment by sindresorhus** [comment]
Looks good :)

**Comment by fisker** [comment]
I just found I might made a mistake, accounding to MSDN https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs#Common_problems data URL don't support query strings

//cc @sindresorhus

**Comment by sindresorhus** [comment]
Would you be able to submit a PR with a fix? We should probably not do anything with the data part, so ignore all logic like query string, hash, path, etc.

**Comment by fisker** [comment]
Yes, working on it

**Pull request merged**